### PR TITLE
Update Eclipse plugin structure, webhooks

### PR DESCRIPTION
**Pull request by niloc132**
In anticipation of the upcoming 4.2 release of the GWT Plugin for Eclipse plus one GWT SDK Plugin release per GWT version, updates the directory structure of plugins.gwtproject.org and the webhooks that should auto-populate these directories.

Manual updates are needed to change the "latest release" and update the composite plugin for new GWT SDK versions.

### CODE REVIEW

File: plugins/content/index.html
Comment (keinhaar) [inline plugins/content/index.html:1]: I expect the new href should point to **gwt-sdk-plugins**?

**Comment by keinhaar** [comment]
My proposal for the new directory structure should be compatible to the current release, because there are only new folders added.
But in your draft impl you placed all folders below the **site** folder. Is that intended?

**Comment by niloc132** [comment]
@keinhaar the one thing that seems missing here is a top-level `site/nightly/` composite, so that users can easily subscribe to the nightly build rather than just releases - what do you think of that idea?

**Comment by niloc132** [comment]
The new `site/` directory is explicitly taken from https://github.com/gwt-plugins/documentation/pull/27#issuecomment-2483849275 - I assumed `repo/` was just a plain top-level directory to enclose everything, but that means that you had two top-level dirs?

Edit: Sorry, I misunderstood, site is the parent of the composite. I'll correct that.

**Comment by niloc132** [comment]
A few questions as I update this to reflect your comment:

`gwt-eclipse-plugin/` is singular, but `gwt-sdk-plugins/` is plural? Are you sure you want the SDK plugin directory to be called plugins? Are we putting more than one plugin in the one directory?

Is there no nice way to let each SDK be downloaded separately and independently, so that we don't need to update the entire SDK plugin with old and new GWT versions each time a new one is changed? Could we add more to the composite, one for each GWT version, and have gwt-sdk-2.10.1/, gwt-sdk-2.11.0, gwt-sdk-2.12.0/ and so on? Then, users install from the composite, and pick and choose for each update which one they want?

We could then easily also add HEAD-SNAPSHOT as its own version that way, and update it nightly, without needing to re-upload each past plugin as well, and users could opt-in to working from the nightly build to better help test future releases.

**Comment by keinhaar** [comment]
> `gwt-eclipse-plugin/` is singular, but `gwt-sdk-plugins/` is plural? Are you sure you want the SDK plugin directory to be called plugins? Are we putting more than one plugin in the one directory?

Depends on the following question ;-)

> Is there no nice way to let each SDK be downloaded separately and independently, so that we don't need to update the entire SDK plugin with old and new GWT versions each time a new one is changed? Could we add more to the composite, one for each GWT version, and have gwt-sdk-2.10.1/, gwt-sdk-2.11.0, gwt-sdk-2.12.0/ and so on? Then, users install from the composite, and pick and choose for each update which one they want?

I've also thought about that, because releases of SDKs would be faster and smaller. A nice way? No. But i would do the extra work if we decide to do so.
The composite can contain more then 2 sites. No Problem here. But we have to update the composite site files with each new SDK version (adding a new repo). Then we would need separate Build processes for each of the SDKs (maybe some subfolders each with it's own pom.xml). We would also need a feature definition per SDK, and each SDK/Feature would need a unique name (GWT-SDK-2.12.1 for example) as well as it's own directory below gwt-sdk-plugins.
Apart from the extra work, this would definitely be the best solution.
Tell me if i should go that way.
 
> We could then easily also add HEAD-SNAPSHOT as its own version that way, and update it nightly, without needing to re-upload each past plugin as well, and users could opt-in to working from the nightly build to better help test future releases.

Agreed.

**Comment by niloc132** [comment]
It is up to you - I'm not an eclipse user, and while I'd prefer to not have approx N^2 copies of GWT releases (say build 1 has 3 SDKs, build 2 has 4 SDKs, build 3 has 5 SDKs...) or some fixed "only the last 4 SDKs in a given version" (still ends up being around 400mb on the server and for each user to download per new GWT release...), that is the current status quo.

I think the real risk is understanding how much of the packaging must change because of new Eclipse versions - do we end up with multiple versions of each GWT SDK for each release? If not, I would sort of hope that we only need to do each GWT release as a one-off (except HEAD-SNAPSHOT of course). There's a lot that I don't understand here.

We should be able to template the xml file or even write/find a bit of code to generate it.

**Comment by niloc132** [comment]
There are a few other things to formalize - should be laid out in a document somewhere under SCM so we can continue to evolve it as things change. Besides the others we've discussed above, one main irritation remains: the artifact name.

When you run the current eclipse plugin build, you get a `repository.zip` file. That is the name of the artifact that is available when the nightly build hook runs, so that it can be consistently downloaded and unzipped on the server. In contrast, some releases presently have artifacts with a name related to the release version, while some keep that consistent `repository.zip`. I would prefer that nightlies and releases have the same deployment where possible - download the zip, unpack into its own directory, etc.

Here's the releases we have across the two repos
```
https://github.com/gwt-plugins/gwt-eclipse-plugin/releases/download/v4.1.0-rc2/gwt_eclise_plugin_v4.1.0_rc2.zip
https://github.com/gwt-plugins/gwt-eclipse-plugin/releases/download/v4.1.0-rc1/gwt_eclise_plugin_v4.1.0_rc1.zip
https://github.com/gwt-plugins/gwt-eclipse-plugin/releases/download/v4.0.0/repository.zip

https://github.com/gwt-plugins/gwt-sdk-plugins/releases/download/v4.1.1-rc1/repository.zip
```
Ideally we won't keep with spelling "eclipse" wrong as a pattern. I have some curl+jq magic to deal with this, but its a little gross to need to make URLs into versions rather than the other way around.

---

After that just one last issue I can see, how do we signal that a given release is now "latest"? This is important not just to know "when a release occurs, should this be marked latest", but also "if we have to re-deploy `gwtproject.org` to new hardware (using this repo's scripts), which do we pick as current?" 
 * We could have a variable and initiate a PR to this repo when it changes
 * We could assume something like "latest version sorted lexigraphically", so that newer bugfix releases of older releases wouldn't replace newer minor releases
 * Just set it manually, since our need to handle new hardware is pretty minimal, and can be managed directly

I have a rough draft of what I think will serve as deployment scripts for initial setup and for webhooks on nightlies and releases of GPE itself, and just need to nail down remaining details on the SDK plugin, and we should be ready to start testing this.

**Comment by niloc132** [comment]
Pushed the updated webhook wiring - confirmed that manually running the init script and the nightly script correctly pulls all the expected builds:
screenshot938
Will have to try the releases live, or tweak it to run on a different repo.

**Comment by niloc132** [comment]
@keinhaar any further thoughts? I can probably deploy this as-is with no issues, but I'd rather finish the discussion, or backburner this for other changes.

**Comment by keinhaar** [comment]
looks good. i would like to wait until i've prepared anything else.
Currently i'm preparing a 4.2 Release of the GWT-Eclipse-Plugin without the SDKs, and the SDKs separated as single Plugins as discussed above. I've changed the versions numbers of each single sdk plugin to match the contained GWT version. 
That would look like this as directory structure.
```
gwt-sdk-plugins
   2.10.1
   2.11.0
   2.12.1
```
Does that make sense for you?
New Versions can be added in just a few seconds, because i've made a little script, that creates the necessary directory structure.
All of this needs some local testing, but it already looks pretty good.

**Comment by niloc132** [comment]
Two thoughts:
 * Take the "s" off of "plugins" - I understand that the repo itself holds more than one plugin at a time, but a user is typically going to move to the next one in line, rather than keep several at once? This isn't a strong opinion - maybe we just rename the directory to be `sdks/` or something?
 * Do we worry that there could ever be a packaging change that would require versioning the "plugin" wrapper of each SDK? That is, what do you think of adding a `-N` on the end:
     * 2.12.0-1
     * 2.12.1-1
     * 2.12.1-2 (if a packaging error or update is required to keep using 2.12.1 in current eclipse versions)
     
   We could use this to go back and release newer versions of the same plugins so that newer eclipse builds can be supported. I guess this probably isn't needed for packaging errors at all, since we would simply want to replace the current SDK plugin with a new one.

Finally I would just need to understand how to get the plugins - would you publish each as its own release?

**Comment by keinhaar** [comment]
> * Take the "s" off of "plugins"

Now, after separating the SDKs completely, it looks logically to me to have this folder named "plugins", because each of them can be installed serarated from the other versions.

> * adding a `-N` on the end

At the moment i don't see any reason for repackaging. If that ever happens, we could still make an 2.12.1-2 folder and link that in the composite site files.

> Finally I would just need to understand how to get the plugins

I could make releases of each, but i believe it would be easier for both of us, if you just create them yourself. Just checkout and run ```mvn clean package``` for each version you will upload to the update site.
The compositeArtifacts.xml and the compositeContent.xml both need to be updated.
compositeContent.xml:
```
<?xml version='1.0' encoding='UTF-8'?>
<?compositeMetadataRepository version='1.0.0'?>
<repository name='GWT Eclipse Plugin'
    type='org.eclipse.equinox.internal.p2.metadata.repository.CompositeMetadataRepository'
    version='1.0.0'>
  <properties size='1'>
    <property name='p2.timestamp' value='1665755219001'/>
  </properties>
  <children size='6'>
    <child location='https://plugins.gwtproject.org/eclipse/gwt-eclipse-plugin/latest'/>
    <child location='https://plugins.gwtproject.org/eclipse/gwt-sdk-plugins/2.8.2'/>
    <child location='https://plugins.gwtproject.org/eclipse/gwt-sdk-plugins/2.9.0'/>
    <child location='https://plugins.gwtproject.org/eclipse/gwt-sdk-plugins/2.10.1'/>
    <child location='https://plugins.gwtproject.org/eclipse/gwt-sdk-plugins/2.11.0'/>
    <child location='https://plugins.gwtproject.org/eclipse/gwt-sdk-plugins/2.12.1'/>
  </children>
</repository>
```

compositeArtifacts.xml:
```
<?xml version='1.0' encoding='UTF-8'?>
<?compositeArtifactRepository version='1.0.0'?>
<repository name='GWT Eclipse Plugin'
    type='org.eclipse.equinox.internal.p2.artifact.repository.CompositeArtifactRepository' 
    version='1.0.0'>
  <properties size='1'>
    <property name='p2.timestamp' value='1665755219001'/>
  </properties>
  <children size='6'>
    <child location='https://plugins.gwtproject.org/eclipse/gwt-eclipse-plugin/latest'/>
    <child location='https://plugins.gwtproject.org/eclipse/gwt-sdk-plugins/2.8.2'/>
    <child location='https://plugins.gwtproject.org/eclipse/gwt-sdk-plugins/2.9.0'/>
    <child location='https://plugins.gwtproject.org/eclipse/gwt-sdk-plugins/2.10.1'/>
    <child location='https://plugins.gwtproject.org/eclipse/gwt-sdk-plugins/2.11.0'/>
    <child location='https://plugins.gwtproject.org/eclipse/gwt-sdk-plugins/2.12.1'/>
  </children>
</repository>
```

**Comment by niloc132** [comment]
Sounds good, I'll rename back to gwt-sdk-plugins/, and update those files. At some point we'll get the webhook to create those files, but for now it probably make sense to let them be manual (along with the "latest" links)?

**Comment by niloc132** [comment]
Please take a look - I'm happy to merge and deploy this now, then once 4.2 is out (and all the applicable sdk versions), we'll do another round of testing and switch to using the new composite url?

**Pull request merged**